proto/private/utils/direct_source_infos.bzl

```diff
@@ -0,0 +1,72 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Contains the implementation for `proto_common.direct_source_infos`."""
+
+ProtoFileInfo = provider(
```

Comment:
Why is this a provider? You probably will not want to return this from a rule since this describes only a single file, and in that case, a `struct()` does the job just as well, but is a bit less confusing.

Reply:
Providers make generating useful documentation easier. The alternative would be to document the structure of the `struct()` in the documentation of `direct_source_infos`'s return type, which, to me, sounds worse than making it a provider.

Reply:
But I un-exported it from `//proto:defs.bzl`. Users won't need to instantiate it.

Reply:
Au contraire! Providers convey the meaning that they are used for communication between configured targets, which isn't the case here. Mind turning it into a struct?

Reply:
Ack, done.